File: app/src/main/java/jp/gr/java_conf/foobar/testmaker/service/views/adapters/TestAdapter.java
package jp.gr.java_conf.foobar.testmaker.service.views.adapters;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.GradientDrawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import net.cattaka.android.adaptertoolbox.adapter.ScrambleAdapter;
import net.cattaka.android.adaptertoolbox.adapter.listener.ForwardingListener;

import jp.gr.java_conf.foobar.testmaker.service.R;
import jp.gr.java_conf.foobar.testmaker.service.models.Test;
import jp.gr.java_conf.foobar.testmaker.service.views.ImageTextButton;

/**
 * Created by keita on 2017/05/21.
 */

public class TestAdapter extends ScrambleAdapter.AbsViewHolderFactory<TestAdapter.ViewHolder>  {

    private Context context;

    public interface OnClickListener {
        void onClickPlayTest(long id);
        void onClickEditTest(long id);
        void onClickDeleteTest(long id);
    }

    @Nullable
    private OnClickListener listener;
    public void setOnClickListener(OnClickListener listener) {
        this.listener = listener;
    }


    public TestAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public TestAdapter.ViewHolder onCreateViewHolder(@NonNull ScrambleAdapter<?> adapter, @NonNull ViewGroup parent, @NonNull ForwardingListener<ScrambleAdapter<?>, RecyclerView.ViewHolder> forwardingListener) {

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_test, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ScrambleAdapter adapter, @NonNull ViewHolder holder,
                                 final int position, Object object) {

        final Test data = (Test)adapter.getItemAt(position);

        holder.title.setText(data.getTitle());

        holder.num.setText(context.getString(R.string.number_existing_questions,data.getQuestionsCorrectCount(),data.getQuestions().size()));

        holder.play.setOnClickListener(view -> {

            if (listener != null) {
                listener.onClickPlayTest(data.getId());
            }

        });

        holder.edit.setOnClickListener(v -> {

            if (listener != null) {
                listener.onClickEditTest(data.getId());
            }
        });

        holder.delete.setOnClickListener(v -> {

            if (listener != null) {
                listener.onClickDeleteTest(data.getId());
            }

        });

        GradientDrawable drawable = (GradientDrawable) context.getResources().getDrawable(R.drawable.circle);

        drawable.setColor(data.getColor());

        holder.cate.setBackgroundDrawable(drawable);

        holder.share.setOnClickListener(view -> {
            Toast.makeText(context, context.getString(R.string.message_share_exam,data.getTitle()), Toast.LENGTH_LONG).show();

            try {
                Intent intent = new Intent();
                intent.setAction(Intent.ACTION_SEND);
                intent.setType("text/plain");

                intent.putExtra(Intent.EXTRA_TEXT, data.testToString(context));
                context.startActivity(intent);

            } catch (Exception e) {
                Log.d("tag", "Error");
            }
        });

    }

    @Override
    public boolean isAssignable(Object object) {
        return object instanceof Test;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView title;
        TextView num;

        ImageTextButton play;
        ImageTextButton edit;
        ImageTextButton delete;
        ImageTextButton share;

        ImageButton cate;

         ViewHolder(View v) {
            super(v);

            num = v.findViewById(R.id.num_questions);
            title = v.findViewById(R.id.title_questions);
            play = v.findViewById(R.id.play);
            edit = v.findViewById(R.id.edit);
            delete =  v.findViewById(R.id.delete);
            share = v.findViewById(R.id.open);
            cate =  v.findViewById(R.id.cate);

        }
    }
}

File: app/src/main/java/jp/gr/java_conf/foobar/testmaker/service/views/adapters/TestAdapter.kt
package jp.gr.java_conf.foobar.testmaker.service.views.adapters

import android.content.Context
import android.content.Intent
import android.graphics.drawable.GradientDrawable
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast

import net.cattaka.android.adaptertoolbox.adapter.ScrambleAdapter
import net.cattaka.android.adaptertoolbox.adapter.listener.ForwardingListener

import jp.gr.java_conf.foobar.testmaker.service.R
import jp.gr.java_conf.foobar.testmaker.service.models.Test
import jp.gr.java_conf.foobar.testmaker.service.views.ImageTextButton

/**
 * Created by keita on 2017/05/21.
 */

class TestAdapter(private val context: Context) : ScrambleAdapter.AbsViewHolderFactory<TestAdapter.ViewHolder>() {

    private var listener: OnClickListener? = null

    interface OnClickListener {
        fun onClickPlayTest(id: Long)
        fun onClickEditTest(id: Long)
        fun onClickDeleteTest(id: Long)
    }

    fun setOnClickListener(listener: OnClickListener) {
        this.listener = listener
    }

    override fun onCreateViewHolder(adapter: ScrambleAdapter<*>, parent: ViewGroup, forwardingListener: ForwardingListener<ScrambleAdapter<*>, RecyclerView.ViewHolder>): TestAdapter.ViewHolder {

        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.card_test, parent, false))
    }

    override fun onBindViewHolder(adapter: ScrambleAdapter<*>, holder: ViewHolder,
                                  position: Int, `object`: Any?) {

        val data = adapter.getItemAt(position) as Test

        holder.title.text = data.title

        holder.num.text = context.getString(R.string.number_existing_questions, data.questionsCorrectCount, data.questions.size)

        holder.play.setOnClickListener {

            if (listener != null) listener!!.onClickPlayTest(data.id)

        }

        holder.edit.setOnClickListener {

            if (listener != null) listener!!.onClickEditTest(data.id)
        }

        holder.delete.setOnClickListener {

            if (listener != null) listener!!.onClickDeleteTest(data.id)

        }

        val drawable = ResourcesCompat.getDrawable(context.resources,R.drawable.circle,null) as GradientDrawable

        drawable.setColor(data.color)

        holder.cate.background = drawable

        holder.share.setOnClickListener {

            Toast.makeText(context, context.getString(R.string.message_share_exam, data.title), Toast.LENGTH_LONG).show()

            try {
                val intent = Intent()
                intent.action = Intent.ACTION_SEND
                intent.type = "text/plain"

                intent.putExtra(Intent.EXTRA_TEXT, data.testToString(context))
                context.startActivity(intent)

            } catch (e: Exception) {
                Log.d("tag", "Error")
            }
        }

    }

    override fun isAssignable(`object`: Any): Boolean {
        return `object` is Test
    }

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {

        var cate: ImageButton = v.findViewById(R.id.cate)
        var title: TextView = v.findViewById(R.id.title_questions)
        var num: TextView = v.findViewById(R.id.num_questions)
        var play: ImageTextButton = v.findViewById(R.id.play)
        var edit: ImageTextButton = v.findViewById(R.id.edit)
        var delete: ImageTextButton = v.findViewById(R.id.delete)
        var share: ImageTextButton = v.findViewById(R.id.open)

    }
}
